include/nektar_interface/geometry_transport_3d.hpp

```diff
@@ -0,0 +1,586 @@
+  /// A local copy of the geometry object.
+  std::shared_ptr<Nektar::SpatialDomains::Geometry3D> geom;
+  /// The underlying Nektar++ shape type
+  LibUtilities::ShapeType shape_type;
+
+  /**
+   * Wrapper around remote 3D geometry object.
+   *
+   * @param rank Owning rank.
+   * @param id Geometry id on remote rank.
+   * @param geom Shared pointer to geometry object.
+   */
+  template <typename T>
+  RemoteGeom3D(const int rank, const int id, std::shared_ptr<T> geom)
+      : rank(rank), id(id),
+        geom(std::dynamic_pointer_cast<Nektar::SpatialDomains::Geometry3D>(
+            geom)),
+        shape_type(geom->GetShapeType()){};
+
+  /**
+   * Get the Nektar++ bounding box for the geometry object.
+   *
+   * @returns Bounding box.
+   */
+  std::array<NekDouble, 6> GetBoundingBox() {
+    return this->geom->GetBoundingBox();
+  }
+};
+
+/**
+ * Get all 3D geometry objects from a Nektar++ MeshGraph
+ *
+ * @param[in] graph MeshGraph instance.
+ * @param[in,out] std::map of Nektar++ Geometry3D pointers.
+ */
+inline void get_all_elements_3d(
+    Nektar::SpatialDomains::MeshGraphSharedPtr &graph,
+    std::map<int, std::shared_ptr<Nektar::SpatialDomains::Geometry3D>> &geoms) {
+  geoms.clear();
+
+  for (auto &e : graph->GetAllTetGeoms()) {
+    geoms[e.first] =
+        std::dynamic_pointer_cast<Nektar::SpatialDomains::Geometry3D>(e.second);
+  }
+  for (auto &e : graph->GetAllPyrGeoms()) {
+    geoms[e.first] =
+        std::dynamic_pointer_cast<Nektar::SpatialDomains::Geometry3D>(e.second);
+  }
+  for (auto &e : graph->GetAllPrismGeoms()) {
+    geoms[e.first] =
+        std::dynamic_pointer_cast<Nektar::SpatialDomains::Geometry3D>(e.second);
+  }
+  for (auto &e : graph->GetAllHexGeoms()) {
+    geoms[e.first] =
+        std::dynamic_pointer_cast<Nektar::SpatialDomains::Geometry3D>(e.second);
+  }
+}
+
+/**
+ * Struct to hold local and remote 3D geoms.
+ */
+struct GeometryLocalRemote3D {
+  /// Local geometry objects which are owned by this MPI rank.
+  std::map<int, std::shared_ptr<Geometry3D>> local;
+  /// Remote geometry objects where a copy is stored on this MPI rank.
+  std::vector<std::shared_ptr<RemoteGeom3D>> remote;
+
+  /**
+   * Push a geometry object onto the container depending on if the geometry
+   * object is local or remote.
+   *
+   * @param geom Geometry object.
+   */
+  inline void push_back(std::pair<int, std::shared_ptr<Geometry3D>> geom) {
+    this->local[geom.first] = geom.second;
+  }
+
+  /**
+   * Push a geometry object onto the container depending on if the geometry
+   * object is local or remote.
+   *
+   * @param geom Geometry object.
+   */
+  inline void push_back(std::shared_ptr<RemoteGeom3D> geom) {
+    this->remote.push_back(geom);
+  }
+
+  /**
+   *  Returns number of remote and local geometry objects.
+   */
+  inline size_t size() { return this->local.size() + this->remote.size(); }
+};
+
+/**
+ * Struct to holder shared pointers to the different types of 3D geometry
```

Comment:
```suggestion
 * Struct to hold shared pointers to the different types of 3D geometry
```
More examples below; suggest find-replace